### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by ikmsk10**
* Initialize meson build scripts
* Supplement README.MD of build and design description.
* Trivial implementation managering Sessions:
  + Create SessionManager
  + Create Session
  + Adjust cleanup callback
  + Remove session by uniquie ID.

TODO:
* Remove session by username, remote IPv4 address, type, unconditional.
* Control unique session grouped by:
  + Username
  + SessionType

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>

### CODE REVIEW

File: src/session.hpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    void adjustSessionOwner(const std::string& userName);
+
+  private:
+    SessionManager::SessionIdentifier identifier;
+    sdbusplus::bus::bus& bus;
+    /** @brief Path of the group instance */
+    const std::string path;
+    SessionManagerWeakPtr managerWeakPtr;
+    std::optional<SessionManager::SessionCleanupFn> cleanupFn;
```
Comment (nest1ing) [inline src/session.hpp]: It seems to me `std::optional` doesn't bring any benefit here but requires one extra byte to keep state.
I don't insist on removing it, but suggest you to reflect on it.

File: src/session.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                                     "' is not found");
+        }
+
+        this->associations({
+            make_tuple("user", "session", userObjectPath),
+        });
+    }
+    catch (const std::exception& ex)
+    {
+        throw std::runtime_error(ex.what());
```
Comment (nest1ing) [inline src/session.cpp]: It looks strange. Likely, you  don't need the `try-catch` statement at all here.

File: README.MD [hunk 3]
```diff
@@ -2,3 +2,73 @@
...
+```sh
+$ source /path/to/sdk/environment-setup-arm1176jzs-openbmc-linux-gnueabi
+$ meson --buildtype plain --optimization s build_dir
+$ ninja -C build_dir
+```
+If build process succeeded, the directory `build_dir` contains executable file
+`libobmcsession.so`.
+
+## Design
+
```
Comment (nest1ing) [inline README.MD]: Do you plan to inject this library into each service that the user communicate with?

File: src/session.cpp (hunk 2 above)
Comment (ikmsk10) [reply src/session.cpp]: I've aimed to feel free of sdbusplus API for `libobmcsession` consumers. I don't desire to include sdbusplus headers only to catch exceptions.

File: src/session.hpp (hunk 1 above)
Comment (ikmsk10) [reply src/session.hpp]: The idea is a cleanup function might be empty to do nothing on the session is closing.
Please, note, the Session has two constructors: with `cleanupFn` and not.

File: README.MD (hunk 3 above)
Comment (ikmsk10) [reply README.MD]: Yes, this is the goal the `libobmcsession` developed on.

File: src/session.hpp (hunk 1 above)
Comment (nest1ing) [reply src/session.hpp]: Yes, and `SessionManager::SessionCleanupFn cleanupFn = nullptr` exactly defines this state.
Additionally, you can check whether `cleanupFn` is defined in this way:
```cpp
if (cleanupFn)
{
    std::invoke(cleanFn, ...);
}
```

File: src/session.cpp (hunk 2 above)
Comment (nest1ing) [reply src/session.cpp]: You are catching it as std::exception and it doesn't require `sdbusplus` API usage.
And one more thing, what() contains the full description of the original exception, including the exception class name. And you are copying it to the new instance. So, I don't see the difference.
It seems to me you can skip this try-catch statement and catch std::exception in the code where you are planing to do it in the same way as you are doing it here.

File: README.MD [hunk 4]
```diff
@@ -2,3 +2,73 @@
...
+|                                                                |                                                              |
+|                      Libobmcsession                            |                          Service 1                           |
+|                                                                |                                                              |
+|                                                                |        +---------------------+                               |
+|                                                                |        |    authorization    +----+                          |
+|                                                                |        +---------------------+    |                          |
+|                                                                |                                   |                          |
+| +-----------------------------------------+                    |                                 +-+------------------------+ |
+| |Publish SessionManager object on the DBus|<----------+        |                                 | internal session staff   | |
```
Comment (AlexanderAmelkin) [inline README.MD]: `staff` -> `stuff`, there is a big difference :)

File: README.MD [hunk 5]
```diff
@@ -2,3 +2,73 @@
...
+| +-----------------------------------------+           |        |                                 +-+------------------------+ |
+|                                                       |        |                                   |                          |
+|       +-----------------------+                       |        |        +--------------------------v------------+             |
+|       | Create session object |<------------+         +--------+--------+ create libobmcsession::SessionManager |             |
+|       +----------+------------+             |                  |        +---------------------------------------+             |
+|                  |                          |                  |                                                              |
+| +----------------v-------------------+      |                  |                                                              |
+| | Publics Session object on the DBus |      |                  |                                                              |
```
Comment (AlexanderAmelkin) [inline README.MD]: `Publics` -> `Publish`

File: README.MD [hunk 6]
```diff
@@ -2,3 +2,73 @@
...
+|        |Adjust session cleanup|<------------+------------------+---------+call libobmcsession::SessionManager::create |       |
+|        +-------^--------------+                                |         +--------------------------------------------+       |
+|                |                                               |                                                              |
+|                |                                               |                                                              |
+|                |                                               |                            ...                               |
+|                |                                               |                                                              |
+|                |                                               |                                                              |
+|                |  +-----------------------+                    |         +-----------------------------------+                |
+|                +--+Remove object from DBus+--------------------+---------+  remove session by some condition |                |
```
Comment (AlexanderAmelkin) [inline README.MD]: I suppose, in the service it's "delete the libobmcsesson::SessionManager object" ?
Also, I suppose that it also calls `xyz.openbmc_project.Object.Delete` under the hood (in the destructor) ?

File: README.MD [hunk 7]
```diff
@@ -2,3 +2,73 @@
...
+|                               |                                |                                                              |
+|                               |                                |                                                              |
+|                               |                                |                                                              |
+|              +----------------+-----------------+              |                                                              |
+|              |Call DBus method                  |              |                                                              |
+|              |xyz.openbmc_project.Object.Delete |              |                                                              |
+|              ++--------------------------------++              |     +-------------------------------------------------+      |
+|               +--------------------------------+               |     |  Remove service_1 session by condition          |      |
+|               |Call SessionManager::remove(...)|<--------------+-----+(username, remoteAddress, SessionId, ofssionType)|      |
```
Comment (AlexanderAmelkin) [inline README.MD]: what is `offsionType` ?

File: include/libobmcsession/manager.hpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    SessionManager(const SessionManager&) = delete;
+    SessionManager& operator=(const SessionManager&) = delete;
+    SessionManager(SessionManager&&) = delete;
+    SessionManager& operator=(SessionManager&&) = delete;
+
+    /** @brief Constructs session manager
+     *
+     * @param[in] bus     - Handle to system dbus
+     * @param[in] objPath - The Dbus service slug uniquely identifies the source
+     *                      of a session to created appropriate items from.
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp]: `created` -> `create the`

File: include/libobmcsession/manager.hpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    ~SessionManager() = default;
+    SessionManager(const SessionManager&) = delete;
+    SessionManager& operator=(const SessionManager&) = delete;
+    SessionManager(SessionManager&&) = delete;
+    SessionManager& operator=(SessionManager&&) = delete;
+
+    /** @brief Constructs session manager
+     *
+     * @param[in] bus     - Handle to system dbus
+     * @param[in] objPath - The Dbus service slug uniquely identifies the source
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp]: `identifies` -> `identifying`

File: include/libobmcsession/manager.hpp [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+     *                      of a session to created appropriate items from.
+     *                      Service name template:
+     *                      'xyz.openbmc_project.Session.${slug}'.
+     * @param[in] type    - Type of all session items which will be created by
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp]: `which` -> `that`, it's a restrictive clause.

https://www.dictionary.com/e/that-vs-which/
https://owl.purdue.edu/owl/general_writing/grammar/that_vs_which.html

File: include/libobmcsession/manager.hpp [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+     *                      the current instance.
+     */
+    SessionManager(sdbusplus::bus::bus& bus, const std::string& slug,
+                   const SessionType type) :
+        bus(bus),
+        slug(slug), serviceName(serviceNameStartSegment + slug), type(type)
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp:56]: is clang-format happy with this fomatting? to me it looks strange.
please borrow the `.clang-format` file from any OpenBMC project and use it.

File: include/libobmcsession/manager.hpp [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    {}
+
+    /**
+     * @brief Create a session and publish into the dbus.
+     *
+     * @param userName - the owner user name
+     * @param remoteAddress - the remote ip-address of session creating
+     * initiators'.
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp]: Align the line, please.
Also, change it it "the IP address of the session initiator", probably that will remove the need for alignment.

File: include/libobmcsession/manager.hpp [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+     * @return SessionIdentifier - uniquie session ID
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp]: unique

File: include/libobmcsession/manager.hpp [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+     */
+    SessionIdentifier create(const std::string& userName,
+                             const uint32_t remoteAddress);
+
+    /**
+     * @brief Create a session and publish into the dbus with cleanup callback
+     * on the session destroy.
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp]: Please align the second line for better readability:

```
@brief Create ...
       on the ...
```

File: include/libobmcsession/manager.hpp [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+     *
+     * @param userName - the owner user name
+     * @param remoteAddress - the remote ip-address of session creating
+     * initiators'.
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp]: same as in the other `create` function

File: include/libobmcsession/manager.hpp [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+     * identifier
+     */
+    const std::string getSessionManagerObjectPath() const;
+
+    /**
+     * @brief Cast session identifier to the hex view.
+     *
+     * @return const std::string - a hex view of session identifier.
+     */
+    static const std::string hexSessionId(SessionIdentifier);
```
Comment (AlexanderAmelkin) [inline include/libobmcsession/manager.hpp:206]: looks like a candidate for an assignment operator override. I don't insist though.

File: src/session.cpp [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+void SessionItem::resetCleanupFn(SessionManager::SessionCleanupFn&& cleanup)
+{
+    this->cleanupFn.reset();
+    this->cleanupFn.emplace(cleanup);
+}
+
+void SessionItem::adjustSessionOwner(uid_t)
+{
+    // Define interface to adjust session owner by user ID.
+    // Currently, not implemented.
```
Comment (AlexanderAmelkin) [inline src/session.cpp]: Why have this stub at all? If it is ever needed, we can add it then.

File: src/session.cpp (hunk 2 above)
Comment (AlexanderAmelkin) [reply src/session.cpp]: I totally agree with @nest1ing. This looks weird.

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: README.MD (hunk 7 above)
Comment (ikmsk10) [reply README.MD]: Something went wrong with converting a diagram to the ASCII view. It should be `sessionType`.

File: src/session.cpp (hunk 17 above)
Comment (ikmsk10) [reply src/session.cpp]: Currently, each `not implemented` method is not required for the next RC, but these must-haves for the feature of the control session by admin role from UI.  I prefer to define the public interface of the shared library to be considered by all library consumers. This is only for saving time, but I can implement those methods for the initial version either.

File: include/libobmcsession/manager.hpp (hunk 11 above)
Comment (ikmsk10) [reply include/libobmcsession/manager.hpp:56]: Yes, I've also been astonished to have seen the same indents. But standard OpenBMC clang-format configuration is already included in the project.

File: src/session.hpp (hunk 1 above)
Comment (ikmsk10) [reply src/session.hpp]: Fixed

File: src/session.cpp (hunk 2 above)
Comment (ikmsk10) [reply src/session.cpp]: Fixed

File: README.MD (hunk 4 above)
Comment (ikmsk10) [reply README.MD]: Fixed

File: README.MD (hunk 5 above)
Comment (ikmsk10) [reply README.MD]: Fixed

File: README.MD (hunk 6 above)
Comment (ikmsk10) [reply README.MD]: Yes, the xyz.openbmc_project.Object.Delete calls under the hood, but this is the internal implementation.
The current diagram displays the public interface only.
The `xyz.openbmc_project.Object.Delete` might be called directly whether through libobmcsession or not.

File: README.MD (hunk 7 above)
Comment (ikmsk10) [reply README.MD]: Fixed

File: include/libobmcsession/manager.hpp (hunk 8 above)
Comment (ikmsk10) [reply include/libobmcsession/manager.hpp]: Fixed

File: include/libobmcsession/manager.hpp (hunk 14 above)
Comment (ikmsk10) [reply include/libobmcsession/manager.hpp]: Fixed

File: include/libobmcsession/manager.hpp (hunk 15 above)
Comment (ikmsk10) [reply include/libobmcsession/manager.hpp]: Fixed

File: include/libobmcsession/manager.hpp (hunk 16 above)
Comment (ikmsk10) [reply include/libobmcsession/manager.hpp:206]: Currently, I've kept it as is.

File: src/session.cpp (hunk 17 above)
Comment (ikmsk10) [reply src/session.cpp]: Fixed

**Review by AlexanderAmelkin — APPROVED** [review]

**Pull request merged**